src/Microsoft.ML.EntryPoints/PermutationFeatureImportance.cs

```diff
@@ -0,0 +1,420 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.ML;
+using Microsoft.ML.CommandLine;
+using Microsoft.ML.Data;
+using Microsoft.ML.EntryPoints;
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Transforms;
+
+[assembly: LoadableClass(typeof(void), typeof(PermutationFeatureImportanceEntryPoints), null, typeof(SignatureEntryPointModule), "PermutationFeatureImportance")]
+
+namespace Microsoft.ML.Transforms
+{
+    internal static class PermutationFeatureImportanceEntryPoints
+    {
+        [TlcModule.EntryPoint(Name = "Transforms.PermutationFeatureImportance", Desc = "Permutation Feature Importance (PFI)", UserName = "PFI", ShortName = "PFI")]
+        public static PermutationFeatureImportanceOutput PermutationFeatureImportance(IHostEnvironment env, PermutationFeatureImportanceArguments input)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            var host = env.Register("Pfi");
+            host.CheckValue(input, nameof(input));
+            EntryPointUtils.CheckInputArgs(host, input);
+
+            var modelOps = new ModelOperationsCatalog(env);
+            var model = modelOps.Load(input.ModelPath.OpenReadStream(), out DataViewSchema schema) as TransformerChain<ITransformer>;
+            // If model.LastTransformer is not an IPredictionTransformer, get the part of the TransformerChain
+            // up to the last ITransformer that is indeed an IPredictionTransformer. This piece of the TransformerChain
+            // is used to extract the IPredictionTransformer and also to transform the input data.
+            // Will throw if there is no IPredictionTransformer in the TransformerChain.
+            if (!(model.LastTransformer is IPredictionTransformer<IPredictorProducing<float>>
+                || model.LastTransformer is IPredictionTransformer<IPredictorProducing<VBuffer<float>>>))
+                model = model.RewindToLastPredictionTransformer();
+            var predictor = model.LastTransformer as ISingleFeaturePredictionTransformer<object>;
+            Contracts.Assert(predictor != null, "Permutation Feature Importance (PFI) is not supported for the predictor.");
+
+            var transformedData = model.Transform(input.Data);
+            IDataView result = PermutationFeatureImportanceUtils.GetMetrics(env, predictor, transformedData, input);
+            return new PermutationFeatureImportanceOutput { Metrics = result };
+        }
+    }
+
+    internal sealed class PermutationFeatureImportanceOutput
+    {
+        [TlcModule.Output(Desc = "The PFI metrics")]
+        public IDataView Metrics;
+    }
+
+    internal sealed class PermutationFeatureImportanceArguments : TransformInputBase
+    {
+        [Argument(ArgumentType.Required, HelpText = "The path to the model file", ShortName = "path", Visibility = ArgumentAttribute.VisibilityType.EntryPointsOnly)]
+        public IFileHandle ModelPath;
```

Comment:
>IFileHandle [](start = 15, length = 11)

In most of the entrypoints, models are passed in and out as objects of type `PredictorModel`. I think it would make more sense to have a `PredictorModel` input here as well, that well the entrypoint can be added to an entrypoint graph.

Reply:
I did try this. The PFI implementation here ([https://github.com/dotnet/machinelearning/blob/edfd10f131abe97a7f7bb25c0070843a67b8b660/src/Microsoft.ML.Transforms/PermutationFeatureImportance.cs#L20-L31](https://github.com/dotnet/machinelearning/blob/edfd10f131abe97a7f7bb25c0070843a67b8b660/src/Microsoft.ML.Transforms/PermutationFeatureImportance.cs#L20-L31)) expects an `IPredictionTransformer`, while `PredictorModel` has an `IPredictor` that cannot be cast to `IPredictionTransformer`.

---
In reply to: [328094891](https://github.com/dotnet/machinelearning/pull/4232#discussion_r328094891) [](ancestors = 328094891)

Reply:
`PredictorModel` has a method called `PrepareData` which gives back the `IPredictor` and a `RoleMappedData` (see an example of how it's used in SummarizePredictor.cs line 41). Then you can use the `IPredictor` and the `RoleMappedSchema` to create the correct `ISingleFeaturePredictionTransformer` as is done in the method `LoadFromLegacy` in TransformerChain.cs line 253.
Having the input be an `IFileHandle` requires two separate entrypoint graphs (like you have in your test), and requires serialization and deserialization of the model.

---
In reply to: [328344099](https://github.com/dotnet/machinelearning/pull/4232#discussion_r328344099) [](ancestors = 328344099,328094891)

Reply:
I see what you mean by adding it to the graph. Thanks for these pointers - I'll look into the code and do another iteration.

---
In reply to: [328499170](https://github.com/dotnet/machinelearning/pull/4232#discussion_r328499170) [](ancestors = 328499170,328344099,328094891)